### PR TITLE
MAINT: upload nighlighties to new location

### PR DESCRIPTION
**Pull request by jarrodmillman**

### CODE REVIEW

**Comment by tupui** [comment]
I am +1 here. Do you plan on sending an email to the mailing list? Happy to facilitate the discussion if needed.

**Review by pllim — COMMENTED** [review]
Won't this be very disruptive to downstream packages already testing against scipy nightly wheel? Should the old location mirror the new one for a while until downstream packages have a chance to catch up?

**Comment by tupui** [comment]
Good point. Since it's nightlies, I would personally push for breaking the current address and tell people to switch to the new one.

**Comment by bsipocz** [comment]
My proposal is that once all the core packages that currently push to the scipy.anaconda space move to the scientific-python.anaconda space, and there is at least one nightly for all of them, we open automated PRs for all downstream we find using the old URL. I volunteer to get these PRs done.

I don't expect anything to break, the old URL will keep working, it's just not getting new versions.

**Comment by martinfleis** [comment]
A bit of context - the [Scientific Python Developer Summit](https://scientific-python.org/summits/developer/2023/) is happening this week. A part of the effort is to ensure that Scientific Python core packages follow the upcoming SPEC documents and support their implementation across the ecosystem. One of the specific changes is a move of the nightly wheels from `pypi.anaconda.org/scipy-wheels-nightly` which is planned to be closed to `pypi.anaconda.org/scientific-python-nightly-wheels`. That is what this PR does.

Once all of the packages that currently use `scipy-wheels-nightly` switch to `scientific-python-nightly-wheels`, we will use a bot to open PRs in downstream repositories to automatically change the index for nightlies across the ecosystem to minimise breakage.

One more step will need to happen here though. The upload requires an upload token stored as a repository secret (as you have right now). However, you will need to generate a new token for the new anaconda organisation. @tupui should already have access and be able to generate a token.

The same change is happening in the other packages as well.

**Comment by andyfaff** [comment]
Is the old package repo going to be switched off at the same time?

**Comment by tupui** [comment]
Still need to run through the mailing list on our side (will do in the next days.) @bsipocz's plan is sensible to me.

**Comment by bsipocz** [comment]
> Is the old package repo going to be switched off at the same time?

No, but afaik the current owner is interested in eventually retiring that one (assuming by repo you mean the https://anaconda.org/scipy-wheels-nightly space)

**Comment by larsoner** [comment]
> I should've been clearer, after the transition we won't want people to be using stale files, so we should remove them from the old location.

Another option would be to push a dummy `scipy 2.0` package to the old location that just `raise RuntimeError('This wheel repo is retired, use https://... instead')` or something. As a repo maintainer I expect these nightly wheels to break every once in a while, and it would be a 1 minute fix for me with such an error message... And if someone is relying / pinned (for now) to a specific nightly version for now (e.g., while working on compat with current nightly) it'll continue to work until we do shut down the old repo.

**Comment by andyfaff** [comment]
I'm all for merging this right away. nightly builds have the potential to break, so it shouldn't be surprising that something breaks.

**Comment by tylerjereddy** [comment]
I don't mind disrupting the downstream consumers for less-stable nightlies, but maybe double check Ralf isn't too annoyed -- there's a nearly-identical discussion over at https://github.com/numpy/numpy/pull/23792, where there is a bit of concern about the amount of disruption for N unknown nightly wheel consumers for a cosmetic name change.

File: doc/source/release/1.7.0-notes.rst [hunk 1]
```diff
@@ -581,7 +581,7 @@ Issues closed for 1.7.0
 * `#13511 <https://github.com/scipy/scipy/issues/13511>`__: DOC: error in description of "direc" parameter of "fmin_powell"
 * `#13526 <https://github.com/scipy/scipy/issues/13526>`__: TST: stats: intermittent \`test_ttest_ind_randperm_alternative2...
 * `#13536 <https://github.com/scipy/scipy/issues/13536>`__: \`_within_tolerance\` seems an unnecessary repetition of \`numpy.isclose\`
-* `#13540 <https://github.com/scipy/scipy/issues/13540>`__: missing python 3.8 manylinux wheels on scipy-wheels-nightly
+* `#13540 <https://github.com/scipy/scipy/issues/13540>`__: missing python 3.8 manylinux wheels on scientific-python-nightly-wheels
```
Comment (bashtage) [inline doc/source/release/1.7.0-notes.rst:584]: Probably shouldn't revisit old release notes. I suspect you did a find/replace.

**Comment by tupui** [comment]
The PR on NumPy's side was merged today. If there are no concerns (nothing so far at least on the mailing list), I would say we should follow them in the very near future.

I propose Friday if nothing comes up in the meantime.

**Comment by tupui** [comment]
@jarrodmillman could you update the linux_meson job to also use the new URL for NumPy? https://anaconda.org/scientific-python-nightly-wheels/numpy

**Comment by rgommers** [comment]
This seems fine to me, folks seem to be either +1 or neutral on this. Let me also link the spec document that was referred to higher up: https://scientific-python.org/specs/spec-0004/. And a similar discussion with a bit more context in https://github.com/scikit-learn/scikit-learn/pull/26417.

> There are now a bunch of PRs to all sorts of projects that rename the location to which nightly wheels are uploaded, but no coordination around making the switch. I think it might be worth taking a step back before merging all these PRs.

This would indeed have been nice (something like "write SPEC including switchover plan, propose change to projects, plan switchover date, open PRs"), but I think the ship has sailed since PRs for NumPy and a few other projects were merged already. So let's do this soon I'd say.

> However, you will need to generate a new token for the new anaconda organisation

Can I please get permissions on the new Anaconda bucket? Username is the same (https://anaconda.org/rgommers).

**Comment by martinfleis** [comment]
> Can I please get permissions on the new Anaconda bucket?

You should have all you need now. The token should need only _Allow write access to the API sit_e and _Allow uploads to Standard Python repositories_ permissions.

**Comment by tylerjereddy** [comment]
I'm removing the milestone because there is really pretty much no connection between this and the release process--nightlies operate off the `main` branch, and you can just add the milestone back if you merge before I branch. This helps me keep the number of PRs on the "review queue" reasonable.

**Comment by rgommers** [comment]
I updated the `SCIPY_NIGHTLY_UPLOAD_TOKEN` token in the secrets of this repo now.

File: doc/source/release/1.7.0-notes.rst (hunk 1 above)
Comment (rgommers) [reply doc/source/release/1.7.0-notes.rst]: I addressed this comment

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks all!

**Comment by tupui** [comment]
@rgommers @andyfaff I am not sure this is working. No package got uploaded so far on the new organization. The last upload on Saturday failed: https://cirrus-ci.com/task/4667447840604160

Do we also need to update `SCIPY_NIGHTLY_UPLOAD_TOKEN` in `cirrus_wheels.yml`?

We do not remove old artifact otherwise right?

---

@rgommers can you please re-create a personal token instead? (It should only have "Allow uploads to Standard Python repositories" and "Allow write access to the API site" scope.)

This is because we reworked a bit the structure of the org and added groups. Only a few folks are left to be admin (still in the work) while the rest should only have access to their group. i.e. people cannot temper with others projects' nightlies. All admin tokens will be revoked soon-ish.

Lastly, since we are using this org now. Can I propose that we endorse the corresponding [SPEC4](https://scientific-python.org/specs/spec-0004/)? (For everyone, endorsing means "we are ok to recommend the practice", not that we do follow exactly the wording in the SPEC.)

**Comment by rgommers** [comment]
It failed with:
```
Using Anaconda API: https://api.anaconda.org/
Using "scientific-python-nightly-wheels" as upload username
Processing './wheelhouse/scipy-1.11.0.dev0-cp311-cp311-macosx_10_9_x86_64.whl'
Detecting file type...
File type is "Standard Python"
Extracting standard python attributes for upload
Creating package "scipy"
Error:  ('User "User object" has not enough permissions to create "User object/scipy" package', 401)
Error: Process completed with exit code 1.
```
probably because my permissions were adjusted before the first scheduled run of the wheel builder job. I have added a new token and manually triggered the CI job. Let's see if that passes, if it does I'll replace the CircleCI token as well.

**Comment by rgommers** [comment]
@tupui same result, I don't have enough permissions:
```
Creating package "scipy"
Error:  ('User "User object" has not enough permissions to create "User object/scipy" package', 401)
```

**Comment by tupui** [comment]
@rgommers I added this edit yesterday on the previous message, not sure you have seen it:

> EDIT: we need to use the admin token first to create the package in the org. Only then we can use a user token to update the package. So we either need to do a manual step or update twice the CI. (Anaconda is really not very good at managing teams...)

Do you still have the admin token around? You need to use it first to create the package because Anaconda does not allow to create empty packages. It's only once a package is created that I can assign it to a group.

**Comment by rgommers** [comment]
> Do you still have the admin token around?

It didn't work, that's the token that was on the job. I'm pretty sure that when you (or whoever did it) revoked my admin permissions before the first scheduled run, that invalidated the token. Please either reinstate the permissions, or manually create a `scipy` package on the bucket.

**Comment by tupui** [comment]
Thanks Ralf 🙇

**Pull request merged**